### PR TITLE
JIT compilation of TODO app with Ivy

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/core/test/bundling/todo/index.ts [hunk 1]
```diff
@@ -145,35 +147,36 @@ export class ToDoAppComponent {
...
-
-// TODO(misko): This hack is here because common is not compiled with Ivy flag turned on.
-(NgIf as any).ngDirectiveDef = defineDirective({
-  type: NgIf,
-  selectors: [['', 'ngIf', '']],
-  factory: () => new NgIf(injectViewContainerRef(), injectTemplateRef()),
-  inputs: {ngIf: 'ngIf', ngIfThen: 'ngIfThen', ngIfElse: 'ngIfElse'}
-});
-
+if (!(NgIf as any).ngDirectiveDef) {
```
Comment (vicb) [inline packages/core/test/bundling/todo/index.ts]: add a comment explaining when this would be the case.

File: packages/core/test/bundling/todo/todo_e2e_spec.ts [hunk 2]
```diff
@@ -26,10 +26,10 @@ describe('functional test for todo', () => {
            const toDoAppComponent = (window as any).toDoAppComponent;
            expect(document.body.textContent).toContain('todos');
            expect(document.body.textContent).toContain('Demonstrate Components');
-           expect(document.body.textContent).toContain('4 items left');
+           expect(document.body.textContent).toMatch(/4\s+items left/);
```
Comment (vicb) [inline packages/core/test/bundling/todo/todo_e2e_spec.ts]: Probably we should just collapse WS ?

File: packages/bazel/src/ng_module.bzl [hunk 3]
```diff
@@ -14,6 +14,90 @@ load(":rules_typescript.bzl",
...
+    fail('unreachable')
+
+def _include_ng_files(ctx):
+  """Determines whether Angular outputs will be produced by the current compilation strategy.
+
+  Args:
+    ctx: skylark rule execution context
+
+  Returns:
+    true iff the current compilation strategy will produce View Engine compilation outputs (such as
```
Comment (trotyl) [inline packages/bazel/src/ng_module.bzl]: iff?

File: packages/bazel/src/ng_module.bzl [hunk 4]
```diff
@@ -429,11 +516,16 @@ ng_module = rule(
...
 
-# TODO(alxhub): this rule exists to allow early testing of the Ivy compiler within angular/angular,
-# and should not be made public. When ng_module() supports Ivy-mode outputs, this rule should be
-# removed and its usages refactored to use ng_module() directly.
-internal_ivy_ng_module = rule(
-    implementation = _ivy_module_impl,
-    attrs = NG_MODULE_RULE_ATTRS,
+
+# TODO(alxhub): this rule causes legacy ngc to produce Ivy outputs from global analysis information.
+# It to facilitate testing of the Ivy runtime until ngtsc is mature enough to be used instead, and
```
Comment (trotyl) [inline packages/bazel/src/ng_module.bzl]: It to?

File: packages/bazel/src/ng_rollup_bundle.bzl [hunk 5]
```diff
@@ -27,6 +27,29 @@ PLUGIN_CONFIG="{sideEffectFreeModules: [\n%s]}" % ",\n".join(
...
+  processing, which affects decorators and annotations.
+
+  In other modes, an emulation of the upstream rollup_bundle rule is used. This avoids running
+  build optimizer on code which isn't designed to be optimized by it.
+
+  Args:
+    ctx: skylark rule execution context
+
+  Returns:
+    true iff the Angular version of rollup with build optimizer should be used, false otherwise
```
Comment (trotyl) [inline packages/bazel/src/ng_rollup_bundle.bzl]: iff?

File: packages/bazel/src/ng_rollup_bundle.bzl [hunk 6]
```diff
@@ -35,7 +58,41 @@ def run_brotli(ctx, input, output):
       arguments = ["--output=%s" % output.path, input.path],
   )
 
+# Borrowed from bazelbuild/rules_nodejs
+def _run_tsc(ctx, input, output):
+  args = ctx.actions.args()
+  args.add(["--target", "es5"])
+  args.add("--allowJS")
```
Comment (trotyl) [inline packages/bazel/src/ng_rollup_bundle.bzl]: `--allowJS` -> `--allowJs`? https://www.typescriptlang.org/docs/handbook/compiler-options.html

File: packages/bazel/src/ng_module.bzl (hunk 4 above)
Comment (mhevery) [reply packages/bazel/src/ng_module.bzl]: maybe be `it is to...`?

File: packages/core/test/bundling/hello_world/BUILD.bazel [hunk 7]
```diff
@@ -44,9 +44,12 @@ jasmine_node_test(
     data = [
         ":bundle",
         ":bundle.js",
-        ":bundle.min.js.br",
+        ":bundle.min.js",
```
Comment (mhevery) [inline packages/core/test/bundling/hello_world/BUILD.bazel]: Why are we removing brotly?

File: packages/core/test/bundling/hello_world/BUILD.bazel [hunk 8]
```diff
@@ -44,9 +44,12 @@ jasmine_node_test(
...
         ":bundle.min_debug.js",
     ],
+    tags = [
+        "ivy-jit",
```
Comment (mhevery) [inline packages/core/test/bundling/hello_world/BUILD.bazel]: Are you saying this this will run in jit only? That is a bit strange, don't you mean not jit?

File: packages/core/test/bundling/hello_world/treeshaking_spec.ts [hunk 9]
```diff
@@ -29,7 +29,8 @@ describe('treeshaking with uglify', () => {
     expect(content).not.toContain('createCommonjsModule');
   });
 
-  it('should not contain zone.js', () => { expect(content).not.toContain('scheduleMicroTask'); });
+  it('should not contain zone.js',
```
Comment (mhevery) [inline packages/core/test/bundling/hello_world/treeshaking_spec.ts]: Why is this change? Seems unrelated?

**Review by mhevery — APPROVED** [review]

File: packages/core/test/bundling/todo/BUILD.bazel [hunk 10]
```diff
@@ -50,6 +51,7 @@ jasmine_node_test(
         ":bundle.min.js",
         ":bundle.min_debug.js",
     ],
+    tags = ["ivy-jit"],
```
Comment (mhevery) [inline packages/core/test/bundling/todo/BUILD.bazel]: So you are saying that this runs in `jit-only`? what about tests for other modes?

File: packages/core/test/bundling/todo/todo_e2e_spec.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/test/bundling/todo/todo_e2e_spec.ts]: Yes, JIT should not preserve WS.

**Review by mhevery — APPROVED** [review]

File: packages/core/test/bundling/todo/todo_e2e_spec.ts (hunk 2 above)
Comment (alxhub) [reply packages/core/test/bundling/todo/todo_e2e_spec.ts]: I agree, but I think we should address it in another PR. I'll capture a task in Asana for now.

File: packages/core/test/bundling/hello_world/BUILD.bazel (hunk 8 above)
Comment (alxhub) [reply packages/core/test/bundling/hello_world/BUILD.bazel]: Not `jit-only` - this tag means the test will be _included_ in the CI environment for `compile=jit`. It'll also run in `compile=legacy` (because `ivy-only` was not specified).

File: packages/bazel/src/ng_module.bzl (hunk 4 above)
Comment (alxhub) [reply packages/bazel/src/ng_module.bzl]: Nice catch, fixed.

File: packages/core/test/bundling/todo/todo_e2e_spec.ts (hunk 2 above)
Comment (alxhub) [reply packages/core/test/bundling/todo/todo_e2e_spec.ts]: I investigated, and it ended up being a one-liner, so I've added whitespace collapsing to this PR.

File: packages/core/src/render3/jit/directive.ts [hunk 11]
```diff
@@ -31,45 +31,65 @@ export function compileComponent(type: Type<any>, metadata: Component): Promise<
...
-  }
+  let def: any = null;
+  Object.defineProperty(type, 'ngComponentDef', {
+    get: () => {
+      if (def === null) {
+        // The ConstantPool is a requirement of the JIT'er.
+        const constantPool = new ConstantPool();
+
+        // Parse the template and check for errors.
+        const template = parseTemplate(templateStr, `ng://${type.name}/template.html`);
```
Comment (vicb) [inline packages/core/src/render3/jit/directive.ts]: Pass the preserveWS option ? 
At least add a TODO ?

File: packages/core/src/render3/jit/directive.ts [hunk 12]
```diff
@@ -31,45 +31,65 @@ export function compileComponent(type: Type<any>, metadata: Component): Promise<
...
   }
+  const templateStr = metadata.template;
 
-  // Parse the template and check for errors.
-  const template = parseTemplate(metadata.template !, `ng://${type.name}/template.html`);
-  if (template.errors !== undefined) {
-    const errors = template.errors.map(err => err.toString()).join(', ');
-    throw new Error(`Errors during JIT compilation of template for ${type.name}: ${errors}`);
-  }
+  let def: any = null;
```
Comment (vicb) [inline packages/core/src/render3/jit/directive.ts]: Could we do better than `any` here ?

**Review by vicb — APPROVED** [review]

File: packages/core/test/bundling/todo/BUILD.bazel (hunk 10 above)
Comment (alxhub) [reply packages/core/test/bundling/todo/BUILD.bazel]: No, it's not marked `ivy-only` so it will run in legacy mode too.

File: packages/core/src/render3/jit/directive.ts (hunk 11 above)
Comment (alxhub) [reply packages/core/src/render3/jit/directive.ts]: Done.

File: packages/core/src/render3/jit/directive.ts (hunk 12 above)
Comment (alxhub) [reply packages/core/src/render3/jit/directive.ts]: Done.
